test/puppetlabs/trapperkeeper/authorization/acl_test.clj

```diff
@@ -78,57 +79,120 @@
 
 (deftest test-global-allow
-  (let [acl (acl/allow "*")]
+  (let [acl (acl/allow {:certname "*"})]
     (testing "should allow anything"
-      (is (acl/allowed? acl "anything")))))
+      (is (acl/allowed? acl {:certname "anything" :extensions {}})))))
 
-(deftest test-acl-matching-with-captures
+(deftest test-acl-certname-matching-with-captures
   (testing "matching backreference of simple name"
-    (let [acl (acl/allow "$1.google.com")]
-      (is (acl/allowed? acl "www.google.com" ["www"]))))
+    (let [acl (acl/allow {:certname "$1.google.com"})]
+      (is (acl/allowed? acl {:certname "www.google.com" :extensions {}} ["www"]))))
   (testing "matching backreference of opaque name"
-    (let [acl (acl/allow "$1")]
-      (is (acl/allowed? acl "c216f41a-f902-4bfb-a222-850dd957bebb" ["c216f41a-f902-4bfb-a222-850dd957bebb"]))))
+    (let [acl (acl/allow {:certname "$1"})]
+      (is (acl/allowed? acl {:certname "c216f41a-f902-4bfb-a222-850dd957bebb"
+                             :extensions {}}
+                        ["c216f41a-f902-4bfb-a222-850dd957bebb"]))))
   (testing "matching backreference of name"
-    (let [acl (acl/allow "$1")]
-      (is (acl/allowed? acl "admin.mgmt.nym1" ["admin.mgmt.nym1"])))))
+    (let [acl (acl/allow {:certname "$1"})]
+      (is (acl/allowed? acl {:certname "admin.mgmt.nym1"
+                             :extensions {}}
+                        ["admin.mgmt.nym1"])))))
+
+(defn challenge
+  ([extensions]
+   (challenge "sylvia.plath.net" extensions))
+  ([certname extensions]
+   (let [default-extensions {:foo "foo"
+                             :baz "baz"
+                             :quux "quux"}]
+     {:certname certname
+      :extensions (merge default-extensions extensions)})))
+
+(deftest test-extension-matching
+  (testing "when using extensions in an ACL"
+    (testing "with scalar matching"
+      (let [allowed? (partial acl/allowed? (acl/allow {:extensions {:tradition "confessional"
+                                                                    :author "plath"}}))
+            !allowed? (comp not allowed?)]
+        (is (allowed? (challenge {:tradition "confessional" :author "plath"})))
+        (is (!allowed? (challenge {:tradition "surrealist" :author "plath"})))
+        (is (!allowed? (challenge {:tradition "surrealist" :author "burroughs"})))
+        (is (!allowed? (challenge {})))
+        (is (!allowed? (challenge {:style "cutup" :publisher "knopf"})))))
+
+    (testing "with list matching"
+      (let [allowed? (partial acl/allowed?
+                              (acl/allow {:extensions {:style ["sonnet" "prose" "sestina"]
+                                                       :author "olds"}}))
+            !allowed? (comp not allowed?)]
+        (is (allowed? (challenge {:style "prose" :author "olds"})))
+        (is (!allowed? (challenge {:style "haiku" :author "olds"})))
+        (is (!allowed? (challenge {:stye "sestina" :author "burroughs"})))))
```

Comment:
Might also be worth testing that `sonnet` and `sestina` are valid styles? Also looks like `stye` was used here instead of `style`, not sure if that's a typo or if that's intentional